src/core/internal/spinlock.d

```diff
@@ -0,0 +1,103 @@
+/**
+ * SpinLock for runtime internal usage.
+ *
+ * Copyright: Copyright Digital Mars 2015 -.
+ * License:   $(WEB www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ * Authors:   Martin Nowak
+ * Source: $(DRUNTIMESRC core/internal/_spinlock.d)
+ */
+module core.internal.spinlock;
+
+import core.atomic, core.thread;
+
+shared struct SpinLock
+{
+    /// for how long is the lock usually contended
+    enum Contention : ubyte
+    {
+        brief,
+        medium,
+        lengthy,
+    }
+
+@trusted @nogc nothrow:
+    this(Contention contention)
+    {
+        this.contention = contention;
+    }
+
+    void lock()
+    {
+        if (cas(&val, size_t(0), size_t(1)))
```

Comment:
I think restructuring this function as a do/while so there's only one cas() might generate tighter code. But not urgent.